clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -243,7 +243,8 @@ public boolean isTransactional() {
 
     @Override
     public boolean isControlBatch() {
-        return (attributes() & CONTROL_FLAG_MASK) > 0;
+        // Due to KAFKA-7467, we need to check for empty control batches and treat them as non-control batches
+        return (attributes() & CONTROL_FLAG_MASK) > 0 && count() > 0;
```

Comment:
We should have a `isEmpty` method instead of doing the count check (in this case, it doesn't matter, but there are many cases where counting is more expensive than checking the presence of one item).